src/backend/query_builder.rs

```diff
@@ -1728,11 +1847,32 @@ pub(crate) fn common_well_known_left_associative(op: &BinOper) -> bool {
     )
 }
 
+macro_rules! join_io {
+    ($iter:ident, $item:ident $(, first $first:expr)?, join $join:expr, do $do:expr $(, last $last:expr)?) => {
+        if let Some($item) = $iter.next() {
+            $($first)?
+            $do
+
+            for $item in $iter {
+                $join
+                $do
+            }
+
+            $($last)?
+        }
+    };
+}
```

Comment:
I was wondering where this macro come from? I think this is brilliant, although may be we should some short examples for it.